File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0103_Binary_Tree_Zigzag_Level_Order_Traversal.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class Question_0103_Binary_Tree_Zigzag_Level_Order_Traversal {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> deque = new LinkedList<>();
        deque.offerLast(root);
        boolean flag = true;
        while (!deque.isEmpty()) {
            int size = deque.size();
            List<Integer> list = new ArrayList<>();
            while (size-- > 0) {
                if (flag) {
                    TreeNode node = deque.pollFirst();
                    list.add(node.val);
                    if (node.left != null) {
                        deque.offerLast(node.left);
                    }
                    if (node.right != null) {
                        deque.offerLast(node.right);
                    }
                } else {
                    TreeNode node = deque.pollLast();
                    list.add(node.val);
                    if (node.right != null) {
                        deque.offerFirst(node.right);
                    }
                    if (node.left != null) {
                        deque.offerFirst(node.left);
                    }
                }
            }
            flag = !flag;
            res.add(list);
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0103_Binary_Tree_Zigzag_Level_Order_Traversal.cc
#include <vector>
#include <deque>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> res;
    if (!root) {
      return res;
    }
    deque<TreeNode*> dq;
    dq.push_back(root);
    bool flag = true;
    while (!dq.empty()) {
      int size = dq.size();
      vector<int> vec;
      while (size--) {
        if (flag) {
          TreeNode* cur = dq.front();
          dq.pop_front();
          if (cur->left) {
            dq.push_back(cur->left);
          }
          if (cur->right) {
            dq.push_back(cur->right);
          }
          vec.push_back(cur->val);
        } else {
          TreeNode* cur = dq.back();
          dq.pop_back();
          if (cur->right) {
            dq.push_front(cur->right);
          }
          if (cur->left) {
            dq.push_front(cur->left);
          }
          vec.push_back(cur->val);
        }
      }
      res.push_back(vec);
      flag = !flag;
    }
    return res;
  }
};
